zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -1052,162 +1092,1030 @@ where
+            })
+        })
+        .collect::<Result<BTreeMap<_, _>, _>>()?;
+
+    let mut tx_extractor = TransactionExtractor::new(finalized);
+    if let Some((spend_vk, output_vk)) = sapling_vk {
+        tx_extractor = tx_extractor.with_sapling(spend_vk, output_vk);
+    }
+    if let Some(orchard_vk) = orchard_vk {
+        tx_extractor = tx_extractor.with_orchard(orchard_vk);
+    }
+    let transaction = tx_extractor.extract()?;
+    let txid = transaction.txid();
+
+    #[allow(clippy::too_many_arguments)]
+    fn to_sent_transaction_output<
+        AccountId: Copy,
+        D: Domain,
+        O: ShieldedOutput<D>,
+        DbT: WalletRead + WalletCommitmentTrees,
+        N,
+    >(
+        domain: D,
+        note: D::Note,
+        output: &O,
+        output_pool: ShieldedProtocol,
+        output_index: usize,
+        pczt_recipient: PcztRecipient<AccountId>,
+        external_address: Option<ZcashAddress>,
+        note_value: impl Fn(&D::Note) -> u64,
+        memo_bytes: impl Fn(&D::Memo) -> &[u8; 512],
+        wallet_note: impl Fn(D::Note) -> Note,
+    ) -> Result<SentTransactionOutput<AccountId>, ExtractErrT<DbT, N>> {
+        let pk_d = D::get_pk_d(&note);
+        let esk = D::derive_esk(&note).expect("notes are post-ZIP 212");
+        let memo = try_output_recovery_with_pkd_esk(&domain, pk_d, esk, output).map(|(_, _, m)| {
+            MemoBytes::from_bytes(memo_bytes(&m)).expect("Memo is the correct length.")
+        });
+
+        let note_value = Zatoshis::try_from(note_value(&note))?;
+        let recipient = match (pczt_recipient, external_address) {
+            (PcztRecipient::External, Some(addr)) => Ok(Recipient::External {
+                recipient_address: addr,
+                output_pool: PoolType::Shielded(output_pool),
+            }),
+            (PcztRecipient::External, None) => Err(PcztError::Invalid(
+                "external recipient needs to have its user_address field set".into(),
+            )),
             #[cfg(feature = "transparent-inputs")]
-            utxos_spent,
+            (PcztRecipient::EphemeralTransparent { .. }, _) => Err(PcztError::Invalid(
+                "shielded output cannot be EphemeralTransparent".into(),
+            )),
+            (PcztRecipient::InternalAccount { receiving_account }, external_address) => {
+                Ok(Recipient::InternalAccount {
+                    receiving_account,
+                    external_address,
+                    note: Box::new(wallet_note(note)),
+                })
+            }
+        }?;
+
+        Ok(SentTransactionOutput::from_parts(
+            output_index,
+            recipient,
+            note_value,
+            memo,
+        ))
+    }
+
+    #[cfg(feature = "orchard")]
+    let orchard_outputs = transaction
+        .orchard_bundle()
+        .map(|bundle| match bundle {
+            OrchardBundle::OrchardVanilla(b) => {
+                assert_eq!(b.actions().len(), orchard_output_info.len());
```

Comment:
let's return `Error` instead of Asset here (x All occurances)

Reply:
at's actually how they do it everywhere (e.g. sapling) and it works well because its lambda. Maybe we should keep their approach in their code?